propolis/src/hw/pci/topology.rs

```diff
@@ -0,0 +1,435 @@
+//! A PCI topology containing one or more PCI buses.
+
+use std::collections::{BTreeMap, BTreeSet};
+use std::io::{Error as IoError, ErrorKind};
+use std::sync::{Arc, Mutex};
+
+use crate::common::RWOp;
+use crate::dispatch::DispCtx;
+use crate::hw::ids;
+use crate::inventory::{Inventory, RegistrationError};
+use crate::mmio::MmioBus;
+use crate::pio::PioBus;
+
+use super::bridge::Bridge;
+use super::{Bdf, Bus, BusLocation, Endpoint, LintrCfg};
+
+use thiserror::Error;
+
+/// A logical identifier for a bus in the topology. A bus's logical identifer
+/// is stable irrespective of the way the topology's bridges are configured.
+#[derive(Clone, Copy, Debug, Ord, PartialOrd, Eq, PartialEq)]
+pub struct LogicalBusId(pub u8);
+
+/// A "routing" identifier for a bus in the topology. The topology considers
+/// bridge configurations when deciding what bus will receive messages directed
+/// using this kind of ID.
+#[derive(Clone, Copy, Ord, PartialOrd, Eq, PartialEq)]
+pub struct RoutedBusId(pub u8);
+
+#[derive(Clone, Copy)]
+struct BusIndex(usize);
+
+/// Errors returned when manipulating PCI topology.
+#[derive(Debug, Error)]
+pub enum PciTopologyError {
+    #[error("The logical bus with ID {0:?} was not found")]
+    LogicalBusNotFound(LogicalBusId),
+
+    #[error("Downstream logical bus ID {0:?} was already registered")]
+    LogicalBusAlreadyExists(LogicalBusId),
+
+    #[error("A PCI device was already attached at {0:?}")]
+    DeviceAlreadyAttached(Bdf),
+
+    #[error("Failed to register a bridge with error {0:?}")]
+    BridgeRegistrationError(RegistrationError),
+}
+
+impl From<PciTopologyError> for IoError {
```

Comment:
There are a couple of places in server/standalone that use it. (They call it because `Instance::initialize` takes a closure that returns an `io::Result`, and errors from this function can bubble up into those init paths; in the server case it's via `MachineInitializer::initialize_chipset`).